config/config.go

```diff
@@ -0,0 +1,264 @@
+
+	return c.createCopy(), nil
+}
+
+func validateNoDictionaryHasValue(rootKey, key Key, cfg Entries) error {
+	subKey, found := splitAtLastOccurrence(key)
+
+	if _, ok := cfg[subKey]; ok && subKey != rootKey {
+		return fmt.Errorf("key %s already has Value set", subKey)
+	}
+
+	if found {
+		return validateNoDictionaryHasValue(rootKey, subKey, cfg)
+	}
+
+	return nil
+}
+
+func splitAtLastOccurrence(s Key) (Key, bool) {
+	// Find the last occurrence of the separator
+	idx := strings.LastIndex(s.String(), keySeparator)
+	if idx == -1 {
+		// If the separator is not found, return the original string and an empty string
+		return s, false
+	}
+	// Split the string at the last occurrence of the separator
+	return s[:idx], true
+}
+
+// Get returns the configuration value for the given key.
+// When the key does not exist false is returned.
+func (c Config) Get(k Key) (Value, bool) {
+	k = sanitizeKey(k)
+
+	v, ok := c.entries[k]
+
+	return v, ok
+}
+
+// GetAll returns a map of all Key-Value-pairs
+func (c Config) GetAll() Entries {
+	return maps.Clone(c.entries)
+}
+
+// GetChangeHistory returns a slice of all changes made to the configuration.
+func (c Config) GetChangeHistory() []Change {
+	return slices.Clone(c.changeHistory)
+}
+
+// Delete removes the configuration Key and Value.
+// Returns a new Config with the Key deleted.
+func (c Config) Delete(k Key) Config {
+	k = sanitizeKey(k)
+
+	for configKey := range c.entries {
+		if configKey == k {
+			delete(c.entries, k)
+			c.changeHistory = append(c.changeHistory, Change{KeyPath: k, Deleted: true})
+		}
+	}
+
+	return c.createCopy()
+}
+
+// DeleteRecursive removes all configuration for the given Key, including all configuration for sub-keys.
+// Returns a new Config with without given Key.
+func (c Config) DeleteRecursive(k Key) Config {
+	k = sanitizeKey(k)
+
+	c.Delete(k)
+
+	//scan for subkeys
+	if k != "" && !strings.HasSuffix(k.String(), keySeparator) {
+		k = k + keySeparator
+	}
+
+	for configKey := range c.entries {
+		if strings.HasPrefix(configKey.String(), k.String()) {
+			delete(c.entries, configKey)
+			c.changeHistory = append(c.changeHistory, Change{KeyPath: configKey, Deleted: true})
+		}
+	}
+
+	return c.createCopy()
+}
+
+// DeleteAll removes all key values pairs for the configuration.
+// Returns a new empty Config with a change history containing all keys that haven been deleted.
+func (c Config) DeleteAll() Config {
+	// delete recursive from root
+	c.DeleteRecursive(keySeparator)
+
+	return Config{
+		entries:            make(Entries),
+		changeHistory:      slices.Clone(c.changeHistory),
+		PersistenceContext: c.PersistenceContext,
+	}
+}
+
+// Diff returns a list of DiffResult with all values that differs for a given key.
+func (c Config) Diff(other Config) []DiffResult {
+	m := make(map[Key]DiffResult, len(c.entries))
+
+	for k, v := range c.entries {
+		m[k] = DiffResult{
+			Key:        k,
+			Value:      v,
+			OtherValue: "",
```

Comment:
Your assumption is correct. An empty string represents that a key is not exist in the `Diff` we cannot distinguish between an empty value or if the key does not exist.
I think it is OK in this use case, because we do need this distinction in our configuration.

Reply:
Hm.. it is hard to say if this will ever lead to any bugs or problems. I think, we should be able to support empty values (in the future) for keys and therefore distinct empty values from not existing ones.

In the blueprint operator we solved this with an exist boolean:
https://github.com/cloudogu/k8s-blueprint-operator/blob/3030786318ef4494c571a46ae3187c98a548dae2/pkg/domain/stateDiffDoguConfig.go#L19

@jelemux suggested a nilpointer as another possible solution. We also experimented with this approach in the blueprint operator for component config, i think. This led to bugs with nilpointers and more code to properly handle this values in `String()` functions.

Reply:
I added a differentiation between missing and empty values in the config-diff